### ISSUE TITLE
a few queries

### ISSUE THREAD

**Issue by oapandit** [ContextualAI/HALOs#5]
Hello Kawin and authors,

Thanks for sharing the excellent work. I enjoyed reading the technical report it was crisp and concise, I appreciate the efforts put in to achieve this clarity.

I have a couple of questions:

1. In the KTO loss function, the second term of expected KL divergence can be interpreted as the average reward score obtained over 'rejected' responses over all input prompts (in desired response case). In this sense, it is similar to DPO where the second term is 'reward' for the rejected response of that particular x but here we are replacing it with the average reward for 'rejected' response from all the xs.
Is this understanding correct?
2.  I saw in the code that there is KTOZero implementation where the expected KL divergence is replaced with 0. I am curious what was its finding.
3. On a lighter note, I would like to know what is the font used in the technical report.

thanks,
Onkar

**Comment by kawin-contextual-ai** [comment]
Thanks for your interest @oapandit !

1. There are two key differences between the KTO and DPO losses. 1) KTO is symmetric; there's no log term that implicitly places more emphasis on negative rewards. 2) The reference point in KTO is the expected reward over all responses for all prompts, which simplifies to the KL term between the policy and reference models; in DPO, the reference is just the rejected response y_rejected.
You are partially correct: in the original KTO implementation, we estimated the KL for the chosen/desirable losses using the rejected/undesirable examples and did the opposite for rejected/undesirable losses. The reason for this is that we wanted to make sure that the gradient with respect to bad generations would always decrease their likelihood, and vice-versa for good generations.
In the latest version of KTO (pushed yesterday), we now estimate the KL term as the average reward for x's and randomly matched outputs y_prime and don't bother backpropagating through the KL term (see [here](https://github.com/ContextualAI/HALOs/blob/44d82bbac74311a6cda0d934b20e7b860df89598/trainers.py#L777)). This means that we just use one KL estimate for both the desirable/undesirable losses and we do not have to have the same number of desirable:undesirable examples in each minibatch. This makes KTO much easier to use: for example, you can throw out 90% of your chosen examples (while keeping all of your rejected examples) and still reach DPO-level performance. ![image](https://github.com/ContextualAI/HALOs/assets/140556241/ec71ea17-46f9-4f8d-b242-4ffd4a13b0f4)

2. KTO-zero is much worse standard KTO, on par with doing another round of SFT.

3. The font is called Libertine.

Let me know if you have any more questions!

**Comment by oapandit** [comment]
Thank you Kawin for answering all the queries, much appreciated!
